### PR TITLE
Make a generic BasicAuthBase trait

### PR DESCRIPTION
**Pull request by er1c**

### CODE REVIEW

**Comment by er1c** [comment]
PR Updated based upon the conversation at lunch, got rid of a isValid(token) signature I had originally added

**Comment by tpunder** [comment]
I think you should rename BasicAuthBase => BasicAuth and the existing BasicAuth to something more like StaticUsersBasicAuth (or something).

Also, we should probably apply a similar refactor to DigestAuth so that it matches.

File: src/main/scala/fm/http/server/BasicAuth.scala [hunk 1]
```diff
@@ -21,41 +21,55 @@ import fm.http.{Headers, Status}
...
+  import BasicAuthBase._
+
+  // Implemented by child classes
+  protected def realm: String
+  protected def isAuthorized(user: String, pass: String): Future[Boolean]
+
   protected def requireAuthImpl(request: Request)(action: => Future[Response]): Future[Response] = {
-    if (isValid(request)) action else {
-      Future.successful(Response(Status.UNAUTHORIZED, Headers(HttpHeaders.Names.WWW_AUTHENTICATE -> s"""Basic realm="$realm"""")))
+    val res: Future[Boolean] = isValid(request)
```
Comment (tpunder) [inline src/main/scala/fm/http/server/BasicAuth.scala]: Remove this line

**Comment by tpunder** [comment]
More thoughts:

It might make sense to hide the StaticUsersBasicAuth within the BasicAuth object and just have it be created using an apply method that maintains source compatibility what how we currently access it.  It could even be an anonymous class.

**Comment by er1c** [comment]
PR Updated for `BasicAuth` changes.  The `DigestAuth` refactoring is more challenging, because we don't have the plaintext passwords as part of our authentication system and we also don't get the the plaintext password from the request header.  Basically we would need to store a digest hash as part of the authentication, which is extra problematic because the digest auth contains the `realm`.

**Pull request merged**